src/torrent.cpp

```diff
@@ -8684,26 +8690,35 @@ namespace {
 		m_ses.deferred_submit_jobs();
 	}
 
-	void torrent::move_storage(std::string const& save_path, move_flags_t const flags)
+	void torrent::move_storage(std::string const& save_path, move_flags_t const flags, std::shared_ptr<std::promise<const storage_moved_alert*>> promise)
 	{
 		TORRENT_ASSERT(is_single_thread());
 		INVARIANT_CHECK;
 
 		if (m_abort)
 		{
-			if (alerts().should_post<storage_moved_failed_alert>())
-				alerts().emplace_alert<storage_moved_failed_alert>(get_handle()
+			if (alerts().should_post<storage_moved_failed_alert>()) {
+				auto* a = alerts().emplace_alert<storage_moved_failed_alert>(get_handle()
 					, boost::asio::error::operation_aborted
 					, "", operation_t::unknown);
+
+				if (promise) promise->set_exception(std::make_exception_ptr(alert_exception(a)));
+			} else {
+				if (promise) promise->set_exception(std::make_exception_ptr(alert_dont_post_exception<storage_moved_failed_alert>()));
+			}
 			return;
 		}
 
 		// if we don't have metadata yet, we don't know anything about the file
 		// structure and we have to assume we don't have any file.
 		if (!valid_metadata())
 		{
-			if (alerts().should_post<storage_moved_alert>())
-				alerts().emplace_alert<storage_moved_alert>(get_handle(), save_path, m_save_path);
+			if (alerts().should_post<storage_moved_alert>()) {
+				auto* a = alerts().emplace_alert<storage_moved_alert>(get_handle(), save_path, m_save_path);
+				if (promise) promise->set_value(a);
```

Comment:
passing the pointer to the alert to the future is risky in the same way as the exception. Making sure that the client gets the future's value before calling `pop_alerts()` is not trivial and, I would expect, a source of errors.